### PR TITLE
[Plan Selection] OpPathTree creation and modification

### PR DESCRIPTION
**Pull request by SimonSapin**
Fixes https://github.com/apollographql/federation-next/issues/128

### CODE REVIEW

File: src/query_graph/path_tree.rs [hunk 1]
```diff
@@ -31,17 +40,263 @@ pub(crate) struct PathTree<TTrigger: Eq + Hash, TEdge: Into<Option<EdgeIndex>>>
...
+                &Arc<TTrigger>,
+                Option<Arc<OpPathTree>>,
+                Vec<(
+                    /* impl Iterator<Item = GraphPathItem<…>> */ _,
+                    &'paths Arc<NormalizedSelectionSet>,
+                )>,
+            )>,
+        > = std::iter::repeat_with(Vec::new)
+            .take(edge_count + 1)
+            .collect();
```
[the remark this answers is not in the document]
Comment (SimonSapin) [reply src/query_graph/path_tree.rs]: The first commit was a direct port. In subsequent commits I’ve refactored to use `IndexMap` in a way that hopefully makes the logic more obvious.

File: src/query_graph/graph_path.rs [hunk 2]
```diff
@@ -189,6 +193,22 @@ pub(crate) struct ClosedBranch(pub(crate) Vec<Arc<ClosedPath>>);
...
+    TEdge: Copy + Into<Option<EdgeIndex>>,
+{
+    pub(crate) fn iter(&self) -> impl Iterator<Item = GraphPathItem<'_, TTrigger, TEdge>> {
+        // FIXME: are these 3 `Vec`s expected to always have the same length?
+        self.edges
+            .iter()
+            .copied()
+            .zip(&self.edge_triggers)
+            .zip(&self.edge_conditions)
+            .map(|((edge, trigger), condition)| (edge, trigger, condition))
```
Comment (SimonSapin) [inline src/query_graph/graph_path.rs:208]: `zip` stops when either input iterator stops and silently truncates the rest, so it’s best with inputs of equal lengths. Is this the case here?

Comment (sachindshinde) [reply src/query_graph/graph_path.rs:208]: Yeah, those three `Vec`s should always be the same length.

File: src/query_graph/path_tree.rs [hunk 3]
```diff
@@ -31,17 +40,244 @@ pub(crate) struct PathTree<TTrigger: Eq + Hash, TEdge: Into<Option<EdgeIndex>>>
...
+    }
+
+    pub(crate) fn from_op_paths(
+        graph: Arc<QueryGraph>,
+        node: NodeIndex,
+        paths: &[(&OpGraphPath, &Arc<NormalizedSelectionSet>)],
+    ) -> Result<Self, FederationError> {
+        assert!(
+            !paths.is_empty(),
+            "Should compute on empty paths" // FIXME: what does this mean?
```
Comment (sachindshinde) [inline src/query_graph/path_tree.rs:79]: It's like a typo, and means "Should not compute on empty paths"

File: src/query_graph/path_tree.rs [hunk 4]
```diff
@@ -31,17 +40,244 @@ pub(crate) struct PathTree<TTrigger: Eq + Hash, TEdge: Into<Option<EdgeIndex>>>
...
+                        conditions: conditions.clone(),
+                        sub_paths_and_selections: vec![(graph_path_iter, selection)],
+                    });
+                }
+            }
+        }
+
+        let childs = merged
+            .into_iter()
+            .map(|(key, value)| {
```
Comment (sachindshinde) [inline src/query_graph/path_tree.rs:155]: The ordering of `Vec` items here will end up being different from the JS code.

The JS code groups items with the same edge (but different triggers) together, in the order in which the triggers were encountered. The JS code also records the order in which edges were encountered for the first time, and orders these same-edge groups according to that.

**Review by sachindshinde — CHANGES_REQUESTED** [review]

File: src/query_graph/path_tree.rs (hunk 4 above)
Comment (SimonSapin) [reply src/query_graph/path_tree.rs]: I’ve done that but with an intermediate struct. Please take a look.

File: src/query_graph/path_tree.rs [hunk 5]
```diff
@@ -31,17 +40,249 @@ pub(crate) struct PathTree<TTrigger: Eq + Hash, TEdge: Into<Option<EdgeIndex>>>
...
+        TEdge: 'inputs,
+    {
+        // Group by and order by unique edge ID, and among those by unique trigger
+        let mut merged = IndexMap::<TEdge, ByUniqueEdge<TTrigger, /* impl Iterator */ _>>::new();
+
+        struct ByUniqueEdge<'inputs, TTriger, GraphPathIter> {
+            target_node: NodeIndex,
+            by_unique_trigger:
+                IndexMap<&'inputs Arc<TTriger>, PathTreeChildInputs<'inputs, GraphPathIter>>,
+        }
```
Comment (sachindshinde) [inline src/query_graph/path_tree.rs:116]: nit: Spelling
```suggestion
        struct ByUniqueEdge<'inputs, TTrigger, GraphPathIter> {
            target_node: NodeIndex,
            by_unique_trigger:
                IndexMap<&'inputs Arc<TTrigger>, PathTreeChildInputs<'inputs, GraphPathIter>>,
        }
```

**Review by sachindshinde — APPROVED** [review]
LGTM, just one minor spelling nit below

**Pull request merged**